modules/bibdk_heimdal/includes/bibdk_heimdal.pages.inc

```diff
@@ -0,0 +1,184 @@
+<?php
+
+/**
+ * @file
+ *
+ * Functions to handle a request for login via a confirmation mail
+ */
+
+module_load_include('inc', 'bibdk_heimdal', 'includes/heimdalDatabase');
+
+/**
+ * Page callback (heimdal/loginrequest/%). Get saved one time login information
+ * from database and act on it
+ *
+ * @param $uuid
+ *   unique identifier from email link
+ */
+function bibdk_heimdal_handle_login_request($uuid){
+  $message = '';
+  $row = array();
+
+  try {
+    $row = heimdalDatabase::get($uuid);
+  }
+  catch (Exception $e){
+    watchdog('heimdal', 'DATABASE EXCEPTION: %message%', array('%message%' => $e->getMessage()), WATCHDOG_ERROR);
+    $message = bibdk_heimdal_exception_message();
+  }
+
+  if(empty($row)){
+    // row was not found - it has probably been deleted because it was too old
+    $message = bibdk_heimdal_link_deprecated_message();
+  }
+
+  if(!empty($row) && !bibdk_heimdal_check_timestamp($row['date'])){
+    // message is to old
+    $message = bibdk_heimdal_link_deprecated_message();
+  }
+
+  $response = bibdk_heimdal_do_insert_culr($row);
+  if(!bibdk_heimcdal_pages_check_insert_response($response)) {
+    // something went wrong when creating culr account
+    $message = bibdk_heimdal_pages_no_culr_message();
+  }
+
+  // row has been fetched - this is a one time link - delete on database
+  heimdalDatabase::deleteByUUID($uuid);
+
+  // if message is set something has gone wrong. Set a message and return to
+  // fromtpage
+  if(!empty($message)){
+    drupal_set_message($message, 'warning');
+    drupal_goto('<front>');
+  }
+
+  $email = $row['email'];
+  if(bibdk_heimdal_pages_create_bibdk_user($email)) {
+    // all is good - set a welcome message and redirect to heimdal access platform
+    // for another login attempt
+    $message = bibdk_heimdal_pages_welcome_message();
+    drupal_set_message($message);
+    $link = bibdk_heimdal_pages_get_login_link();
+    drupal_goto($link);
+  }
+
+  // user was not created on openuserinfo - set a  message and return to front
+  drupal_set_message(bibdk_heimdal_pages_something_went_wrong_message(), 'error');
+  drupal_goto('<front>');
+}
+
+/**
+ * Create a user on openuserinfo - if user doesn't already exist
+ * @param string $mail
+ *   Username
+ */
+function bibdk_heimdal_pages_create_bibdk_user($mail) {
+  if (!BibdkUser::instance()->verify($mail)) {
+    // user does not exist - create
+    // make a random password
+    $pass = user_password();
+    return bibdk_provider_user_create_account($mail, $pass);
+  }
+  return TRUE;
+}
+
+function bibdk_heimdal_pages_something_went_wrong_message(){
```

Comment:
This function and the two below could be joined into one with the text as an input param.